src/Jeffijoe.MessageFormat/Parsing/LiteralParser.cs

```diff
@@ -46,135 +45,145 @@ public IEnumerable<Literal> ParseLiterals(StringBuilder sb)
                     {
-                        matchTextBuf.Append(nextChar);
-                        insideEscapeSequence = true;
-                        currentEscapeSequenceLineNumber = lineNumber;
-                        currentEscapeSequenceColumnNumber = columnNumber;
-                        ++i;
-                    }
-
-                    continue;
-                }
-
-                if (insideEscapeSequence)
-                {
-                    matchTextBuf.Append(c);
-                    continue;
-                }
+                        openBraces++;
+                        braceBalance++;
 
-                if (c == OpenBrace)
-                {
-                    openBraces++;
-                    braceBalance++;
+                        // Record starting position of possible new brace match.
+                        if (braceBalance == 1)
+                        {
+                            start = i;
+                            startColumnNumber = columnNumber;
+                            startLineNumber = lineNumber;
+                            matchTextBuf.Clear();
+                        }
+                    }
 
-                    // Record starting position of possible new brace match.
-                    if (braceBalance == 1)
+                    if (c == CloseBrace)
                     {
-                        start = i;
-                        startColumnNumber = columnNumber;
-                        startLineNumber = lineNumber;
-                        matchTextBuf.Clear();
+                        closeBraces++;
+                        braceBalance--;
+
+                        // Write the brace to the match buffer if it's not the closing brace
+                        // we are looking for.
+                        if (braceBalance > 0)
+                        {
+                            matchTextBuf.Append(c);
+                        }
                     }
-                }
+                    else
+                    {
+                        if (i > start && braceBalance > 0)
+                        {
+                            matchTextBuf.Append(c);
+                        }
 
-                if (c == CloseBrace)
-                {
-                    closeBraces++;
-                    braceBalance--;
+                        continue;
+                    }
 
-                    // Write the brace to the match buffer if it's not the closing brace
-                    // we are looking for.
-                    if (braceBalance > 0)
+                    if (openBraces != closeBraces)
                     {
-                        matchTextBuf.Append(c);
+                        continue;
                     }
+
+                    // Passing in the text buffer instead of the actual string to avoid allocating a new string.
+                    result.Add(new Literal(start, i, startLineNumber, startColumnNumber, matchTextBuf.ToString()));
```

Comment:
I wonder if `ReadOnlyMemory` is better?

Reply:
The ReadOnlyMemory would still need to point to some underlying data, so we need some kind of data store (StringBuilder previously, string now). I think that it will be the same